File: go/register/register_test.go
package register_test

import (
	"bufio"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/nats-io/nats-server/v2/server"
	"github.com/nats-io/nats.go"

	"foldlab/register"
)

type action struct {
	Kind    string `json:"kind"`
	Holder  string `json:"holder"`
	Token   uint64 `json:"token"`
	Outcome string `json:"outcome"`
}

type traceState struct {
	Fields     register.State  `json:"fields"`
	Index      int             `json:"index"`
	Transition json.RawMessage `json:"transition"`
}

type row struct {
	ID    string `json:"id"`
	Trace struct {
		States []traceState `json:"states"`
	} `json:"trace"`
	Attempt  action         `json:"attempt"`
	Verdict  string         `json:"verdict"`
	Law      string         `json:"law"`
	Observed register.State `json:"observed"`
}

func corpus(t *testing.T) []row {
	t.Helper()
	path := filepath.Join("..", "..", "packages", "plait", "fixtures", "register-traces.ndjson")
	f, err := os.Open(path)
	if err != nil {
		t.Fatal(err)
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	if !scanner.Scan() {
		t.Fatal("missing provenance header")
	}
	var header struct {
		Provenance string `json:"provenance"`
		Rows       int    `json:"rows"`
	}
	if err := json.Unmarshal(scanner.Bytes(), &header); err != nil {
		t.Fatal(err)
	}
	if header.Rows != 15 {
		t.Fatalf("corpus count = %d, want 15", header.Rows)
	}
	var rows []row
	for scanner.Scan() {
		var r row
		if err := json.Unmarshal(scanner.Bytes(), &r); err != nil {
			t.Fatal(err)
		}
		rows = append(rows, r)
	}
	if err := scanner.Err(); err != nil {
		t.Fatal(err)
	}
	if len(rows) != header.Rows {
		t.Fatalf("decoded %d rows, want %d", len(rows), header.Rows)
	}
	return rows
}

func start(t *testing.T) (*register.Client, nats.JetStreamContext) {
	t.Helper()
	srv, err := server.NewServer(&server.Options{
		ServerName: "plait-register-test", JetStream: true, StoreDir: t.TempDir(),
		DontListen: true, NoLog: true, NoSigs: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	go srv.Start()
	if !srv.ReadyForConnections(10 * time.Second) {
		t.Fatal("embedded NATS v2.14.4 did not become ready")
	}
	t.Cleanup(func() { srv.Shutdown(); srv.WaitForShutdown() })
	nc, err := nats.Connect("", nats.InProcessServer(srv))
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(nc.Close)
	js, err := nc.JetStream()
	if err != nil {
		t.Fatal(err)
	}
	client, err := register.Open(nc)
	if err != nil {
		t.Fatal(err)
	}
	return client, js
}

func invoke(c *register.Client, work string, a action) (register.State, error) {
	switch a.Kind {
	case "grant":
		return c.Grant(work, a.Holder)
	case "renew":
		return c.Renew(work, a.Token)
	case "commit":
		return c.Commit(work, a.Token, a.Outcome)
	case "expire-steal":
		return c.ExpireSteal(work, a.Holder)
	case "observe":
		return c.Observe(work)
	default:
		return register.State{}, errors.New("unknown action " + a.Kind)
	}
}

func decodeTransition(t *testing.T, raw json.RawMessage) action {
	t.Helper()
	var a action
	if err := json.Unmarshal(raw, &a); err != nil {
		t.Fatal(err)
	}
	return a
}

func TestVeilCorpusReplayWall(t *testing.T) {
	rows := corpus(t)
	replayed := 0
	for _, r := range rows {
		r := r
		t.Run(r.ID, func(t *testing.T) {
			client, _ := start(t)
			const work = "0123456789abcdef"
			for _, step := range r.Trace.States[1:] {
				got, err := invoke(client, work, decodeTransition(t, step.Transition))
				if err != nil {
					t.Fatalf("prefix: %v", err)
				}
				if !statesEqual(got, step.Fields) {
					t.Fatalf("prefix state = %#v, want %#v", got, step.Fields)
				}
			}
			got, err := invoke(client, work, r.Attempt)
			if (err == nil) != (r.Verdict == "accepted") {
				t.Fatalf("verdict err=%v, want %s", err, r.Verdict)
			}
			if err != nil {
				var refusal *register.Refusal
				if !errors.As(err, &refusal) || refusal.Law != r.Law {
					t.Fatalf("refusal = %v, want law %q", err, r.Law)
				}
				got, err = client.Observe(work)
				if err != nil {
					t.Fatal(err)
				}
			}
			if !statesEqual(got, r.Observed) {
				t.Fatalf("observed = %#v, want %#v", got, r.Observed)
			}
			history, err := client.History(work)
			if err != nil && !errors.Is(err, nats.ErrKeyNotFound) {
				t.Fatal(err)
			}
			landed := 0
			for _, entry := range history {
				var value struct {
					Outcome *register.Outcome `json:"outcome"`
				}
				if err := json.Unmarshal(entry.Value(), &value); err != nil {
					t.Fatal(err)
				}
				if value.Outcome != nil {
					landed++
					if value.Outcome.Value == "zombie" {
						t.Fatal("zombie outcome landed")
					}
				}
			}
			if landed > 1 {
				t.Fatalf("landed outcomes = %d, want <= 1", landed)
			}
		})
		replayed++
	}
	if replayed != 15 {
		t.Fatalf("replayed %d rows, want 15", replayed)
	}
}

func statesEqual(a, b register.State) bool {
	ab, _ := json.Marshal(a)
	bb, _ := json.Marshal(b)
	return string(ab) == string(bb)
}

// requireWrongLastSequence asserts the frozen DEV-704 classification: every
// wrong-last-sequence refusal carries API code 10071, and the operation
// context picks the sentinel (Create matches ErrKeyExists; Update wraps
// ErrKeyRevisionMismatch while still matching ErrKeyExists).
func requireWrongLastSequence(t *testing.T, err error, sentinel error, context string) {
	t.Helper()
	if !errors.Is(err, sentinel) {
		t.Fatalf("%s error = %v, want %v", context, err, sentinel)
	}
	var apiErr *nats.APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("%s error = %v, want a JetStream APIError in the chain", context, err)
	}
	if apiErr.ErrorCode != nats.JSErrCodeStreamWrongLastSequence {
		t.Fatalf("%s API code = %d, want %d", context, apiErr.ErrorCode, nats.JSErrCodeStreamWrongLastSequence)
	}
}

func TestSubstrateCASProbes(t *testing.T) {
	_, js := start(t)
	kv, err := js.KeyValue(register.Bucket)
	if err != nil {
		t.Fatal(err)
	}
	first, err := kv.Create("probe", []byte("one"))
	if err != nil {
		t.Fatal(err)
	}
	_, err = kv.Create("probe", []byte("two"))
	if err == nil {
		t.Fatal("duplicate create succeeded")
	}
	requireWrongLastSequence(t, err, nats.ErrKeyExists, "duplicate create")
	second, err := kv.Update("probe", []byte("two"), first)
	if err != nil || second <= first {
		t.Fatalf("fresh update = (%d, %v), want larger token", second, err)
	}
	_, err = kv.Update("probe", []byte("zombie"), first)
	if err == nil {
		t.Fatal("stale update succeeded")
	}
	requireWrongLastSequence(t, err, nats.ErrKeyRevisionMismatch, "stale update")
}

File: packages/plait/test/Register.test.ts
import { afterAll, beforeAll, describe, expect, test } from "bun:test"
import { join, resolve } from "node:path"

import { Kvm } from "@nats-io/kv"
import { connect } from "@nats-io/transport-node"
import { Effect, Result, Schema } from "effect"

import { commitWithoutTokenGuard } from "../negative-controls/stale-token-mutant.js"
import {
  REGISTER_BUCKET,
  REGISTER_HISTORY,
  Registers,
  hold,
  type RegisterService,
  type RegisterState,
} from "../src/Register.js"
import { structuralRefusal } from "../src/Refusal.js"
import { startNatsHarness, type NatsHarness, waitForFile } from "./NatsHarness.js"

const Outcome = Schema.Struct({ token: Schema.Number, value: Schema.String })
const State = Schema.Struct({
  token: Schema.Number,
  holder: Schema.NullOr(Schema.String),
  outcome: Schema.NullOr(Outcome),
})
const Action = Schema.Union([
  Schema.Struct({ kind: Schema.Literal("grant"), holder: Schema.String }),
  Schema.Struct({ kind: Schema.Literal("renew"), token: Schema.Number }),
  Schema.Struct({
    kind: Schema.Literal("commit"),
    token: Schema.Number,
    outcome: Schema.String,
  }),
  Schema.Struct({ kind: Schema.Literal("expire-steal"), holder: Schema.String }),
  Schema.Struct({ kind: Schema.Literal("observe") }),
])
type Action = typeof Action.Type
const TraceState = Schema.Struct({
  fields: State,
  index: Schema.Number,
  transition: Schema.Union([Schema.Literal("after_init"), Action]),
})
const Row = Schema.Struct({
  id: Schema.String,
  trace: Schema.Struct({ states: Schema.Array(TraceState), theory: Schema.String }),
  attempt: Action,
  verdict: Schema.Literals(["accepted", "refused"]),
  law: Schema.String,
  observed: State,
})
type Row = typeof Row.Type
const Header = Schema.Struct({ provenance: Schema.String, rows: Schema.Number })

const fixture = resolve(import.meta.dir, "../fixtures/register-traces.ndjson")
let harness: NatsHarness

const loadCorpus = async (): Promise<ReadonlyArray<Row>> => {
  const lines = (await Bun.file(fixture).text()).trimEnd().split("\n")
  const header = Schema.decodeUnknownSync(Header)(JSON.parse(lines[0]!), {
    onExcessProperty: "error",
  })
  expect(header.provenance).toBe(
    "lake exe fabric_veil_export --write-corpus packages/plait/fixtures/register-traces.ndjson",
  )
  const rows = lines.slice(1).map((line) => Schema.decodeUnknownSync(Row)(JSON.parse(line), {
    onExcessProperty: "error",
  }))
  expect(rows).toHaveLength(header.rows)
  expect(rows).toHaveLength(12)
  return rows
}

const invoke = (
  registers: RegisterService,
  work: string,
  action: Action,
): Effect.Effect<RegisterState, import("../src/Refusal.js").Refusal> => {
  switch (action.kind) {
    case "grant": return registers.grant(work, action.holder)
    case "renew": return registers.renew(work, action.token)
    case "commit": return registers.commit(work, action.token, action.outcome)
    case "expire-steal": return registers.expireSteal(work, action.holder)
    case "observe": return registers.observe(work)
  }
}

const runRow = (row: Row) => Effect.gen(function* () {
  const registers = yield* Registers
  const work = "0123456789abcdef"
  for (const step of row.trace.states.slice(1)) {
    if (step.transition === "after_init") throw new Error("after_init may only be the first state")
    const state = yield* invoke(registers, work, step.transition)
    expect(state).toEqual(step.fields)
  }

  const attempted = yield* Effect.result(invoke(registers, work, row.attempt))
  expect(Result.isSuccess(attempted)).toBe(row.verdict === "accepted")
  if (Result.isSuccess(attempted)) {
    expect(attempted.success).toEqual(row.observed)
  } else {
    expect(attempted.failure.law).toBe(row.law)
    expect(yield* registers.observe(work)).toEqual(row.observed)
  }
})

const auditAndDestroy = async (): Promise<void> => {
  const connection = await connect({ servers: harness.url })
  try {
    const bucket = await new Kvm(connection).open(REGISTER_BUCKET)
    const status = await bucket.status()
    expect(status.history).toBe(REGISTER_HISTORY)
    expect(status.ttl).toBe(0)
    expect(status.max_bytes).toBe(-1)
    expect(status.replicas).toBe(1)
    const history = await bucket.history({ key: "0123456789abcdef" })
    let landed = 0
    for await (const entry of history) {
      const value = entry.json<{ readonly outcome: null | { readonly value: string } }>()
      if (value.outcome !== null) {
        landed++
        expect(value.outcome.value).not.toBe("zombie")
      }
    }
    expect(landed).toBeLessThanOrEqual(1)
    expect(await bucket.destroy()).toBe(true)
  } finally {
    await connection.close()
  }
}

beforeAll(async () => {
  harness = await startNatsHarness()
})

afterAll(async () => {
  await harness.stop()
})

describe("Veil register replay wall", () => {
  test("TS equals all 12 model verdicts and observed states with zero skips", async () => {
    const rows = await loadCorpus()
    let replayed = 0
    for (const row of rows) {
      await Effect.runPromise(runRow(row).pipe(
        Effect.provide(Registers.layer({ servers: harness.url })),
        Effect.scoped,
      ))
      await auditAndDestroy()
      replayed++
    }
    expect(replayed).toBe(12)
  }, 120_000)

  test("the substrate refuses duplicate create and stale revision update", async () => {
    const connection = await connect({ servers: harness.url })
    try {
      const bucket = await new Kvm(connection).create(REGISTER_BUCKET, {
        history: REGISTER_HISTORY,
        replicas: 1,
        ttl: 0,
        max_bytes: -1,
      })
      const first = await bucket.create("probe", "one")
      await expect(bucket.create("probe", "two")).rejects.toBeDefined()
      const second = await bucket.update("probe", "two", first)
      expect(second).toBeGreaterThan(first)
      await expect(bucket.update("probe", "zombie", first)).rejects.toBeDefined()
      expect(await bucket.destroy()).toBe(true)
    } finally {
      await connection.close()
    }
  }, 30_000)

  test("the named stale-token mutant is killed by zombie-stale-commit", async () => {
    const row = (await loadCorpus()).find(({ id }) => id === "zombie-stale-commit")
    expect(row).toBeDefined()
    const mutant = commitWithoutTokenGuard(row!.observed, 1, "zombie")
    expect(row!.verdict).toBe("refused")
    expect(mutant).not.toEqual(row!.observed)
  })

  test("heartbeat lease loss interrupts the scoped holder fiber", async () => {
    const lost = structuralRefusal({
      kind: "stale-register-token",
      law: "renew requires the current fencing token",
      path: ["token"],
      got: 1,
      expected: 2,
      next: [],
    })
    let renewals = 0
    const fixture: RegisterService = {
      grant: () => Effect.succeed({ token: 1, holder: "holder", outcome: null }),
      renew: () => {
        renewals++
        return Effect.fail(lost)
      },
      commit: () => Effect.die("unused"),
      expireSteal: () => Effect.die("unused"),
      observe: () => Effect.die("unused"),
    }
    const result = await Effect.runPromise(Effect.result(
      hold("work", "holder", () => Effect.never, "1 millis").pipe(
        Effect.provide(Registers.testLayer(fixture)),
        Effect.scoped,
      ),
    ))
    expect(Result.isFailure(result)).toBe(true)
    if (Result.isFailure(result)) expect(result.failure.law).toBe(lost.law)
    expect(renewals).toBe(1)
  })

  test("a killed TS holder is stolen by Go and its TS zombie is fenced", async () => {
    const schedule = Schema.decodeUnknownSync(Schema.Struct({
      corpusRow: Schema.String,
      steps: Schema.Array(Schema.String),
    }))(JSON.parse(await Bun.file(resolve(import.meta.dir, "../fixtures/crash-steal-schedule.json")).text()))
    const modelRow = (await loadCorpus()).find(({ id }) => id === schedule.corpusRow)
    expect(modelRow?.id).toBe("zombie-stale-commit")
    expect(schedule.steps).toHaveLength(5)

    const work = "fedcba9876543210"
    const readyPath = join(harness.directory, "register-holder.json")
    const zombiePath = join(harness.directory, "register-zombie.json")
    const holder = Bun.spawn({
      cmd: ["bun", "run", "./test/process/register-holder.ts", harness.url, work, readyPath],
      cwd: resolve(import.meta.dir, ".."),
      stdout: "pipe",
      stderr: "pipe",
    })
    await waitForFile(readyPath)
    const ready = Schema.decodeUnknownSync(Schema.Struct({ token: Schema.Number }))(
      JSON.parse(await Bun.file(readyPath).text()),
    )
    holder.kill()
    await holder.exited

    const goRoot = resolve(import.meta.dir, "../../../go")
    const steal = Bun.spawn({
      cmd: ["go", "run", "./cmd/registerwall", harness.url, "steal", work, "go-contender"],
      cwd: goRoot,
      stdout: "pipe",
      stderr: "pipe",
    })
    const stolenText = steal.stdout instanceof ReadableStream
      ? await new Response(steal.stdout).text()
      : ""
    const stealError = steal.stderr instanceof ReadableStream
      ? await new Response(steal.stderr).text()
      : ""
    expect(await steal.exited, stealError).toBe(0)
    const stolen = Schema.decodeUnknownSync(State)(JSON.parse(stolenText))
    expect(stolen.token).toBeGreaterThan(ready.token)
    expect(stolen.holder).toBe("go-contender")

    const zombie = Bun.spawn({
      cmd: [
        "bun", "run", "./test/process/register-zombie.ts",
        harness.url, work, String(ready.token), zombiePath,
      ],
      cwd: resolve(import.meta.dir, ".."),
      stdout: "pipe",
      stderr: "pipe",
    })
    await waitForFile(zombiePath)
    const zombieError = zombie.stderr instanceof ReadableStream
      ? await new Response(zombie.stderr).text()
      : ""
    expect(await zombie.exited, zombieError).toBe(0)
    const refusal = Schema.decodeUnknownSync(Schema.Struct({
      kind: Schema.String,
      law: Schema.String,
    }))(JSON.parse(await Bun.file(zombiePath).text()))
    expect(refusal.kind).toBe("stale-register-token")
    expect(refusal.law).toBe("no stale token ever lands")

    const winner = Bun.spawn({
      cmd: [
        "go", "run", "./cmd/registerwall", harness.url, "commit",
        work, String(stolen.token), "winner",
      ],
      cwd: goRoot,
      stdout: "pipe",
      stderr: "pipe",
    })
    const winnerText = winner.stdout instanceof ReadableStream
      ? await new Response(winner.stdout).text()
      : ""
    const winnerError = winner.stderr instanceof ReadableStream
      ? await new Response(winner.stderr).text()
      : ""
    expect(await winner.exited, winnerError).toBe(0)
    const landed = Schema.decodeUnknownSync(State)(JSON.parse(winnerText))
    expect(landed.outcome).toEqual({ token: stolen.token, value: "winner" })
  }, 120_000)
})
